### PR TITLE
Don't use symbols in `Crystal::Lexer#check_macro_opening_keyword`

### PR DESCRIPTION
**Pull request by HertzDevil**

### CODE REVIEW

File: src/compiler/crystal/syntax/lexer.cr [hunk 1]
```diff
@@ -1966,7 +1966,7 @@ module Crystal
           elsif !delimiter_state && whitespace && (keyword = lookahead { check_macro_opening_keyword(beginning_of_line) })
             char = current_char
 
-            nest += 1 unless keyword == :abstract_def
+            nest += 1 unless keyword == {Keyword::ABSTRACT, Keyword::DEF}
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/lexer.cr:1969]: Considering that this seems to be the only use of `check_macro_opening_keyword`, wouldn't it be sufficient to return `true`/`false` depending on whether it found `abstract def`? 🤔

Comment (HertzDevil) [reply src/compiler/crystal/syntax/lexer.cr:1969]: The truthiness is also used, there'll need to be at least 3 distinct return values

Comment (HertzDevil) [reply src/compiler/crystal/syntax/lexer.cr:1969]: There might be a way

**Comment by Sija** [comment]
What's the point of replacing symbols with named tuples?

**Review by straight-shoota — APPROVED** [review]

**Review by straight-shoota — COMMENTED** [review]
Actually, after thinking about this again I'm not convinced about using a named tuple here just to signal three states.
IMO `Bool?` would be fine. Maybe not ideal for mapping the different values. Perhaps semantics would improve by renaming the method to `keyword_at_macro_start_is_abstract_def?`: `nil` means there is no keyword at macro start, `false` means the keyword is not `abstract def` and `true` means it is.
More fancy modeling is possible but unnecessary for the current usage.

**Comment by HertzDevil** [comment]
`false` would interfere with the way `#lookahead` works (any falsey value causes backtracking, and the other uses of `#lookahead` do return `false`), so there still need to be two distinct truthy states. I was originally going for a private enum at first but it seems even more overkill than the `NamedTuple` here.

**Comment by straight-shoota** [comment]
Thanks for clarification. Makes sense.

I'm not happy about using a named tuple, though. We're usually preaching their purpose is for named arguments and it's not recommended for anything else.
Maybe we can use a record instead? That's basically the same except that we need a line to define the type. But that's all.

**Comment by beta-ziliani** [comment]
I disagree with the idea that NamedTuple is only for named args 😅 I think of NamedTuple as an anonymous record, and I don't see a problem with it.

**Comment by straight-shoota** [comment]
@beta-ziliani I don't think we're using named tuples anywhere like that (and IMO for good reason). Instead, we have dozends of simple `record` types in the compiler source. We should do that here as well.

**Comment by straight-shoota** [comment]
I figure a `record` might be more flexible should there ever be a need for additional information.
But either is fine by me.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**